Rust: Replace rust_crate_type with rust_abi

xclaesse opened the pull request:
-    Rust: Fix both_libraries() case
    
    Rustc does not produce object files we can reuse to build both
    libraries. Ideally this should be done with a single target that has
    both `--crate-type` arguments instead of having 2 different build rules.
    
    As temporary workaround, build twice and ensure they don't get conflicts
    in intermediary files created by rustc by passing target's private
    directory as --out-dir.
    
    See https://github.com/rust-lang/rust/issues/111083.

-    Rust: Use Popen_safe_logged in sanity checks

-    Rust: Fix proc-macro usage when cross compiling
    
    Force BUILD machine and allow to link with HOST machine.

-    Rust: Prevent linking Rust ABI with C library/executable

-    Rust: Remove unit test already covered in "rust/4 polyglot"

-    Rust: Replace rust_crate_type with rust_abi
    
    Meson already knows if it's a shared or static library, user only need
    to specify the ABI (Rust or C).

-    Rust: Add a rust.proc_macro() method

-    interpreter: Use common definition for sources type

-    interpreter: Allow regex matching in expect_error()

xclaesse commented:
CC @sdroege @dcbaker

That's another way to fix https://gitlab.freedesktop.org/slomo/gst-plugin-meson-test/-/commit/78fc0417d28eb129702cab91964fdb98ec521fb8

xclaesse commented:
I'm wondering if we should even deprecate `cdylib`, `dylib`, `staticlib` and `rlib` values. Meson only needs to know which ABI (rust or C) and already knows dynamic vs static.

sdroege commented:
That sounds like an option, just means that things don't map 1:1 to rustc (or cargo) terminology anymore and more documentation would be needed.

xclaesse commented:
While at it, I fixed usage of both_libraries and default_library=both, and added unittest.

> That sounds like an option, just means that things don't map 1:1 to rustc (or cargo) terminology anymore and more documentation would be needed.

Yes, but the alternative is everyone is going to write `library(..., rust_crate_type: 'cdylib')` because it works by default, just like you did in your experiment, and that will have to be fixed when someone wants e.g. --default-library=static. But deprecating those values can be done later, after @dcbaker lands its type checking that will make that easier afaik.

sdroege commented:
I guess that makes sense, just make sure the deprecation warning makes it clear what should be done instead :)

sdroege commented on docs/yaml/functions/_build_target_base.yaml:
```diff
@@ -292,16 +292,17 @@ kwargs:
-      If it is a [[static_library]] it defaults to "lib", and may be "lib",
-      "staticlib", or "rlib". If "lib" then Rustc will pick a default, "staticlib"
-      means a C ABI library, "rlib" means a Rust ABI.
+      If it is a [[static_library]] it defaults to "lib", and may be "lib", "clib"
+      "staticlib", or "rlib". "clib" and "staticlib" means a C ABI library,
```
```suggestion
      If it is a [[static_library]] it defaults to "lib", and may be "lib", "clib",
      "staticlib", or "rlib". "clib" and "staticlib" means a C ABI library,
```

sdroege commented on docs/yaml/functions/_build_target_base.yaml:
```diff
@@ -292,16 +292,17 @@ kwargs:
-      "proc-macro" is a special rust procedural macro crate.
+      If it is a [[shared_library]] it defaults to "lib", and may be "lib", "clib",
+      "dylib", "cdylib", or "proc-macro". "clib" and "cdylib" means a C ABI library,
+      "lib" and "dylib" means a Rust ABI, and "proc-macro" is a special rust
```
```suggestion
      "lib" and "dylib" means a Rust ABI, and "proc-macro" is a special Rust
```

Should probably `s/rust/Rust/` everywhere at some point

sdroege commented on docs/yaml/functions/_build_target_base.yaml:
```diff
@@ -292,16 +292,17 @@ kwargs:
+      procedural macro crate.
 
       "proc-macro" is new in 0.62.0.
+      "clib" is new in 1.2.0.
```
Maybe explain what `clib` / `lib` are actually doing, as in the switch between the shared/static variant as needed (also `lib` is new or not?).

sdroege commented on mesonbuild/backend/ninjabackend.py:
```diff
@@ -1924,7 +1915,8 @@ def generate_rust_target(self, target: build.BuildTarget) -> None:
         # Rustc replaces - with _. spaces or dots are not allowed, so we replace them with underscores
-        args += ['--crate-name', target.name.replace('-', '_').replace(' ', '_').replace('.', '_')]
+        crate_name = self._get_rust_dependency_name(target, target.name).replace(' ', '_').replace('.', '_')
+        args += ['--crate-name', crate_name]
```
This is wrong. You don't want to pass this crate's name through the `rust_dependency_map`. That will even break things if you want to e.g. link a crate with the same name into this crate (then you'd have to rename the dependency via the map but not this crate itself).

xclaesse replied:
The reason I did that is because in my unit test I build the `stuff` crate with different library names: https://github.com/mesonbuild/meson/pull/11714/files#diff-fd68d1be0fd651fd3add1a7ca498172f28c5111ab59ed4b45be1bb8d3a365491R32.

It feels like it should be allowed to have a crate name not matching the lib name, after all that's pretty much the same thing for dependencies. Cargo does not allow that?

sdroege replied:
It allows lib/crate name to be different from the package name (see e.g. `gst-plugin-rtp` as package name vs. `gstrsrtp` as lib/crate name), but in meson terminology that would map to `library` vs. `declare_dependency`.

What you're doing here will disallow certain patterns used by existing crates, most importantly being able to link an older version of the same crate (and giving it a different name locally). I'm using that for example [here](https://github.com/sdroege/ebur128/blob/1b1541c615fe750f7ac56f3f6cfc27b2c2adb53c/Cargo.toml#L26).

You might want a different feature for what you're doing in the test there :)

sdroege replied:
Also quite simply, it makes no sense in cargo to locally change this crate's crate name. It's not observable (apart from logging / backtraces).

For your test you probably want to simply change the `library` name directly instead of such things?

xclaesse replied:
I see... so the alternative is to have a separate `rust_crate_name: 'foobar'`, but is there a use-case for that other than my unit test? If not, I guess the alternative is to generate `stuff.rs` in my unit test with a different `#![crate_name = "stuff"]` line for each test, right?

sdroege replied:
I'm not sure why this is even needed. Couldn't you simply pass a different name to `library()` / `build_target()` each time? That is 1:1 mapped to the crate name.

Maybe I'm missing something here :)

xclaesse replied:
Removed this commit and adapted the unit test for now.

sdroege replied:
Why is it not sufficient to change the target name though?

Apart from that this looks all good to me.

xclaesse replied:
The crate name in stuff.rs is always "stuff" and I cannot build multiple libraries that all have the same name "stuff". Meson complains that multiple targets have the same name because the test matrix build for example a shared library with `rust_crate_type: 'dylib'` and another identical but with `rust_crate_type: 'lib'`.

sdroege replied:
Can't you do `library('stuff_{crate_type}', 'stuff.rs', rust_crate_type: crate_type, ...)`?

xclaesse replied:
That fails at build time:
```
FAILED: libstuff-clib-both_libraries.a 
rustc -C linker=cc --color=always -C debug-assertions=yes --crate-type staticlib -g -C relocation-model=pic --crate-name stuff_clib_both_libraries --emit dep-info=stuff-clib-both_libraries.d --emit link -o libstuff-clib-both_libraries.a ../stuff.rs
error: `--crate-name` and `#[crate_name]` are required to match, but `stuff_clib_both_libraries` != `stuff`
 --> ../stuff.rs:1:1
  |
1 | #![crate_name = "stuff"]
  | ^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to previous error
```

xclaesse replied:
That means that `library('stuff', ...)` and `#![crate_name = "stuff"]` must match exactly. We can't have a lib name different to crate name.

sdroege replied:
Yeah that makes sense. But why do you have `#![crate_name = "stuff"]` in the code? Without it would work.

xclaesse replied:
I was already there, I'm just extending an existing unit test. I did not know that's optional.

sdroege replied:
Ah I see. I'm not exactly sure why it even exists. The answer is probably somewhere in [here](https://doc.rust-lang.org/reference/linkage.html). But for the test, just remove it :)

Also interesting from that link:
> In one session of compilation, the compiler can generate multiple artifacts through the usage of either command line flags [...]

You could build (c)dylib and staticlib/rlib at the same time with the same compiler invocation if you want. Maybe something for a future PR though.

xclaesse replied:
So choice would be lib/clib/both? Yes, let's ignore that for now.

xclaesse replied:
I updated unit test to remove `#![crate_name = "stuff"]` and it works fine indeed. Thanks.

sdroege commented on mesonbuild/build.py:
```diff
@@ -1990,13 +2005,15 @@ def post_init(self) -> None:
                 mlog.debug('Defaulting Rust static library target crate type to rlib')
                 self.rust_crate_type = 'rlib'
+            elif self.rust_crate_type == 'clib':
+                self.rust_crate_type = 'staticlib'
```
For `lib` it's logging something about defaulting, maybe same here?

xclaesse replied:
done

sdroege commented on mesonbuild/build.py:
```diff
@@ -2084,13 +2092,15 @@ def __init__(
                 mlog.debug('Defaulting Rust dynamic library target crate type to "dylib"')
                 self.rust_crate_type = 'dylib'
+            elif self.rust_crate_type == 'clib':
+                self.rust_crate_type = 'cdylib'
```
For `lib` it's logging something about defaulting above, maybe same here?

sdroege commented on test cases/rust/4 polyglot/meson.build:
```diff
@@ -4,6 +4,40 @@ if host_machine.system() == 'darwin'
-l = shared_library('stuff', 'stuff.rs', rust_crate_type: 'cdylib', install : true)
-e = executable('prog', 'prog.c', link_with : l, install : true)
-test('polyglottest', e)
+# Test all combinaisons of crate and target types
```
```suggestion
# Test all combinations of crate and target types
```

xclaesse commented:
@sdroege any idea what could be causing that CI failure? https://github.com/mesonbuild/meson/actions/runs/4789026887/jobs/8516401795?pr=11714#step:4:2263

sdroege commented:
> @sdroege any idea what could be causing that CI failure? https://github.com/mesonbuild/meson/actions/runs/4789026887/jobs/8516401795?pr=11714#step:4:2263

Exciting. I have no idea :) Can you reproduce it locally?

xclaesse commented:
> Exciting. I have no idea :) Can you reproduce it locally?

no, works fine here, maybe more recent gcc and/or rustc.

There is also that failure: https://github.com/rust-lang/cargo/issues/8239. I guess we need -pthread, but it does not seems to be needed locally, so maybe also caused by a different rustc version.

sdroege commented:
You could try with a newer version, I guess. There's some code in ninjabackend.py that is only enabled with 1.66 and before that it's luck if things work correctly in certain configurations or not (and you're running here with 1.65)

xclaesse commented:
And finally on Windows there are other issues too: https://dev.azure.com/jussi0947/jussi/_build/results?buildId=17314&view=logs&j=5015ea34-1c35-577b-b206-c8bea8465a58&t=00c5cf7c-b41c-5970-0bf5-8751e484378e&l=3289

sdroege commented:
On Windows that's missing to link to `ws2_32.dll`, not sure where the linker should get that from here and why it usually works. You found an interesting testcase :)

dcbaker commented:
I had thought about this too... Instead of `rust_crate_type` though, maybe something more generic like `abi_format`? Rust really isn't the only language that we could have this problem with. If we ever got around to adding Zig we would have this problem. D actually has a "BetterC" mode where you write D code, but the compiler disables a bunch of features to ensure C ABI compatibility with no external D runtime usage, which might be a another use case for this.

Then we could move `proc_macro` into the Rust module instead of as a normal target, since it has a bunch of special restrictions that doesn't exist for other crate types.

Just throwing out dieas.

xclaesse commented:
@dcbaker while I like the idea of having something more generic, I don't think we are there yet. We can rethink then when/if we actually do zig and/or D.

sdroege commented on test cases/rust/4 polyglot/meson.build:
```diff
@@ -4,6 +4,46 @@ if host_machine.system() == 'darwin'
-e = executable('prog', 'prog.c', link_with : l, install : true)
-test('polyglottest', e)
+extra_deps = [
+  dependency('threads'),
```
Also `-ldl` apparently. Theoretically we can collect all these by passing `--print=native-static-libs` to `rustc` when linking the `staticlib`

xclaesse replied:
What's weird is it works fine locally. Wondering if rustc now add them by default? In any case, CI should be fixed now.

sdroege commented:
As discussed on IRC, meson should probably make use of the output of `--print=native-static-libs` when building `staticlib` crates. Those libs should also end up in the pkg-config file under `Libs.private`, for example. AFAIU it is required to pass that to the actual build (and not to an empty `dummy.rs` file) as it will print the libs that are actually used by the dependencies built into the crate, and not just the ones pulled in by `std`.

I guess this PR as it is could be merged first and then the above can be handled in a separate PR, but hardcoding the libraries that rustc might pull in is fragile and can break with any newer version if it ends up using a different/additional system library.

xclaesse commented:
CI passed, but I'm worried because I had to restart it to pass. Seems to hit randomly that issue: https://github.com/rust-lang/rust/issues/55931.

xclaesse commented:
I made a minimal test-case to showcase the rustc bug and reported the issue https://github.com/rust-lang/rust/issues/111083.

xclaesse commented:
> I made a minimal test-case to showcase the rustc bug and reported the issue [rust-lang/rust#111083](https://github.com/rust-lang/rust/issues/111083).

After discussion with rustc dev, I found a workaround that should make CI stable.

sdroege commented:
That workaround looks good, even as something to keep for isolating build results more nicely.

Did you check if the resulting link dirs for the dependencies are still correct though? E.g. does the test from https://github.com/mesonbuild/meson/pull/11719 still work afterwards?

xclaesse commented:
> That workaround looks good, even as something to keep for isolating build results more nicely.

Yes, and if you add `-Dsave-temps` to debug all files nicely goes into the target's private dir as they should. I think that's the proper way of doing it for meson even if rustc fix their temp file collision.

> Did you check if the resulting link dirs for the dependencies are still correct though? E.g. does the test from #11719 still work afterwards?

I did it now and it still pass after rebasing that PR on top of this one. Note that Meson specific `lib` and `clib` values are only exposed to user. Internally Meson will always have one of the Rust specific values. That means stuff like `indirect_dep.rust_crate_type in {'staticlib', 'cdylib'}` is still correct in the backend.

sdroege commented:
Yeah that part is clear :) I was just wondering if the change with the directory had any unintended side-effects

xclaesse commented:
@dcbaker @eli-schwartz any objection?

dcbaker reviewed (changes requested):
I haven't looked at all of the code yet, but right now `rust_crate_type` as a 1:1 mapping with what the `--crate-type ` rustc option and the `crate-type` field in cargo.toml. After this series all of those values are deprecated, and all that's left is what ABI to expose. I don't think it makes sense to have `rust_crate_type` that wants no value that `--crate-type` takes, and only values that it doesn't. That's going to be confusing to anyone coming from Rust/Cargo.

At this point let's just not fix `rust_crate_type`, just document it's broken in all kinds of ways, deprecate it, and add an `export_abi` or similar flag which can be used by languages/compilers that support such a thing (off the top of my head D and Rust are the only languages with explicit options for this that Meson supports). This also makes it simpler to map rust_crate_type values to expose_abi values, the same way I changed gui_app and win_subsystem to be reduced in the interpreter instead of the backend.

xclaesse commented:
`rust_crate_type` is also used for `proc-macro`. I'm a bit hesitant to make a generic API such as `export_abi` is we are not sure about any other use-case yet.

dcbaker commented:
@sdroege and I had talked about moving `proc-macro` support into the rust module to avoid the potential of building a host only proc-macro. I'd be willing to write those pathces.

How about `rust_abi` instead of `export_abi`? If I ever get around to adding D support we can just make `rust_abi` an alias of `export_abi` in the interpreter.

xclaesse commented:
> How about `rust_abi` instead of `export_abi`?

TBH that feels like naming nitpicking, I don't really see what's wrong in keeping `rust_crate_type`. It's true that it's not translate 1:1 to rustc --crate-type argument, but it's related, and it's the natural name that people will look into the doc.

> @sdroege and I had talked about moving `proc-macro` support into the rust module to avoid the potential of building a host only proc-macro.

We have a PR for that already: https://github.com/mesonbuild/meson/pull/11743

dcbaker commented:
#11743 is the quick fix of "assert that `if rust_crate_type == 'proc-macro' then for_machine == 'build'`, I proposed there adding `rust_mod.proc_macro()` and deprecating `rust_crate_type == 'proc-macro'` altogether.

I agree that's exactly where someone coming from Rust/Cargo is going to look. I just think it's really confusing to document that "all of the values that are valid to pass to rustc/cargo are invalid, use these other values instead"; and less confusing is: "this has been deprecated in favor of the simpler and more robust `rust_abi` option. That option works correctly for `libary()`, `both_library()` and `build_target()`"

xclaesse commented:
> #11743 is the quick fix of "assert that `if rust_crate_type == 'proc-macro' then for_machine == 'build'`, I proposed there adding `rust_mod.proc_macro()` and deprecating `rust_crate_type == 'proc-macro'` altogether.

It sets `native: true` by default for proc-macro, I'm not sure what else a proc_macro() method would do. Let's cross that bridge when we get there.

dcbaker commented:
Here's a quick and basic working example of what I was thinking: https://github.com/dcbaker/meson/tree/submit/rust-module-proc-macro.

Basically just does some extra error checking and enforces that the crate type is proc-macro, and the the machine is the build machine.

dcbaker commented:
I think this is a real user experience issue, and our prior art is to not add values that not make sense to an existing argument, but to fully deprecate and replace. case and point is `gui_app`, which wasn't extended to take string values, it was deprecated and replaced with `win_subsystem`, which takes a new set of values instead of bools. That made sense. replacing `rust_crate_type` with `rust_abi` makes sense.

dcbaker commented:
> I'm not convinced it's useful to be honest. It just adds a new method for little benefice.

it doesn't make sense to allow users to set arguments that will always error. Setting `rust_crate_type : 'proc-macro', native : false` will always error. Always. providing a method that doesn't allow that, plus can give more useful errors seems like a net win.

xclaesse commented:
@dcbaker well... then this PR is blocked on you making rust.proc_macro() PR. Can you please open one?

xclaesse commented:
I personally still believe it's overkill to add a method for something that really adds no value, but since @sdroege "liked" comments above, I guess that makes 2 vs 1, I'm overruled :P

xclaesse commented:
@dcbaker I added comments on https://github.com/mesonbuild/meson/commit/c0928a3516af624d3cded0175cf0e4c98e870c5c, not sure you get notified if it's not a PR?

sdroege commented:
> @dcbaker Should rust_abi be string or boolean? There is only 2 possible values: Rust or C, right? Maybe rust_abi should default to True and rust_abi: false means C ?

There will very likely be more in the future

xclaesse commented:
@dcbaker @sdroege I included the new rust.proc_macro() method in this PR. It is getting big and I really would like to get things moving forward, could you please do a review asap?

dcbaker reviewed (approved):
Thanks for doing this, and thank you for bearing with me on the keyword argument name. This looks great

eli-schwartz commented:
The https://github.com/nvinson/refivar project no longer builds in meson from git master:

```
meson.build:113:5: ERROR: Invalid rust_crate_type: must be "bin" for executables.
```

The failing code:
```meson
lib_refivar = static_library(
  'efivar',

[...]
  rust_crate_type: 'lib',
[...]
)

rust.test('unit tests', lib_refivar, rust_args:[])
```

Bisecting points to here:
```
10dcd87d002f6f36b1f60371dc807b8d9959d97b is the first bad commit
commit 10dcd87d002f6f36b1f60371dc807b8d9959d97b
Author: Xavier Claessens <xavier.claessens@collabora.com>
Date:   Fri Apr 21 16:21:48 2023 -0400

    Rust: Replace rust_crate_type with rust_abi
    
    Meson already knows if it's a shared or static library, user only need
    to specify the ABI (Rust or C).

 docs/markdown/snippets/rust_crate_type.md   | 11 ++++
 docs/yaml/functions/_build_target_base.yaml |  5 ++
 docs/yaml/functions/library.yaml            | 11 ++++
 docs/yaml/functions/shared_library.yaml     |  8 +++
 docs/yaml/functions/shared_module.yaml      |  8 +++
 docs/yaml/functions/static_library.yaml     |  8 +++
 mesonbuild/backend/ninjabackend.py          | 11 +---
 mesonbuild/build.py                         | 78 +++++++++++++-----------
 mesonbuild/interpreter/type_checking.py     | 25 ++++++++
 test cases/rust/4 polyglot/meson.build      | 51 +++++++++++++++-
 test cases/rust/4 polyglot/proc.rs          |  7 +++
 test cases/rust/4 polyglot/stuff.rs         |  2 -
 test cases/rust/4 polyglot/test.json        | 94 +++++++++++++++++++++++++++--
 13 files changed, 263 insertions(+), 56 deletions(-)
 create mode 100644 docs/markdown/snippets/rust_crate_type.md
 create mode 100644 test cases/rust/4 polyglot/proc.rs
```

/cc @nvinson

dcbaker commented:
@eli-schwartz, @xclaesse: https://github.com/mesonbuild/meson/pull/12383
